# <p style="text-align:center"> **GIAI ĐOẠN 2: TIỀN XỬ LÝ DỮ LIỆU** </p>
---
- **Về công cụ**: Sử dụng numpy và pandas để tiền xử lý dữ liệu. 
- Bước này dùng để tiền xử lý dữ liệu, task có sẵn ở bên dưới.
---

### <p style="text-align:center"> **I. IMPORT THƯ VIỆN CẦN THIẾT** </p>

In [119]:
# Import thư viện cần thiết để tiền xử lý dữ liệu
import pandas as pd
import numpy as np 

### <p style="text-align:center"> **II. DATA EXPLORATIONS AND PREPROCESSING** </p>
#### **0. Cái nhìn sơ lược về bộ dữ liệu**
- Bộ dữ liệu được chia làm 4 phần, tương ứng với mỗi phần là vị trí của cầu thủ trong một đội bóng. 4 vị trí lần lượt là defender (hậu vệ), forward (tiền đạo), keeper (thủ môn), midfielder (tiền vệ). 

In [120]:
# Đọc hết dữ liệu trong CSV về các biến tương ứng
dfKeeper = pd.read_csv('../DATA/keeper_data.csv')
dfForward = pd.read_csv('../DATA/forward_data.csv')
dfDefender = pd.read_csv('../DATA/defender_data.csv')
dfMidfielder = pd.read_csv('../DATA/midfielder_data.csv')

- Để có cái nhìn sơ lược về dữ liệu, nhóm em in ra 5 dòng đầu tiên và các đặc điểm chính của mỗi bộ dữ liệu.

In [121]:
dfKeeper.head() 

Player name  Appearances  Clean sheets  Wins  Losses Saves  \
0             Adrián          139            38    57      45   464   
1            Alisson          150            68   104      17   341   
2   Álvaro Fernández           12             1     3       7    49   
3    Alphonse Aréola           39             9     5      21   118   
4  Kepa Arrizabalaga           86            29    45      24   186   

   Penalties saved  Punches  High claims  Catches  ...  \
0                5       57          150       53  ...   
1                2       67           64       21  ...   
2                0        4            8        4  ...   
3                0       27           23        5  ...   
4                1       25           38       12  ...   

   Errors leading to goal Own goals Yellow cards  Red cards  Fouls  Goals  \
0                      10         0           10          2      5      0   
1                       8         1            2          1      3      1   
2                       1         1            0          0      0      0   
3                       2         0            2          0      1      0   
4                       3         0            3          0      0      0   

   Assists  Passes  Passes per match  Accurate long balls  
0        0   3,481             25.04                1,096  
1        3   4,471             29.81                  617  
2        0     456             38.00                   80  
3        0     913             23.41                  177  
4        0   2,435             28.31                  398  

[5 rows x 25 columns]

In [122]:
dfForward.head()

Player name  Appearances  Goals Overview  Wins  Losses  Goals  \
0    Terry Ablade            0               0     0       0      0   
1   Tammy Abraham           89              26    36      33     26   
2  Adam Armstrong           52               3     8      32      3   
3    Adama Traoré          172               9    49      76      9   
4   Simon Adingra            0               0     0       0      0   

   Goals per match  Headed goals  Goals with right foot  Goals with left foot  \
0             0.00             0                      0                     0   
1             0.29             3                     20                     1   
2             0.06             0                      1                     2   
3             0.05             1                      7                     1   
4             0.00             0                      0                     0   

   ...  Crosses  Yellow cards Red cards  Fouls Offsides  Tackles  \
0  ...        0             0         0      0        0        0   
1  ...       32             2         0     56       39       25   
2  ...       39             2         0     28        4       24   
3  ...      547             8         0    142       30      122   
4  ...        0             0         0      0        0        0   

   Blocked shots  Interceptions Clearances  Headed clearance  
0              0              0          0                 0  
1             31             12         55                45  
2             19              6         10                 7  
3             55             34         40                22  
4              0              0          0                 0  

[5 rows x 31 columns]

In [123]:
dfDefender.head()

Player name  Appearances  Goals Overview  Wins  Losses  Clean sheets  \
0       Zach Abbott            0               0     0       0             0   
1  Tosin Adarabioyo           43               1     9      19            10   
2    Tayo Adaramola            0               0     0       0             0   
3      Joshua Addae            0               0     0       0             0   
4      Nayef Aguerd            1               0     0       1             0   

   Goals conceded  Tackles Tackle success  Last man tackles  ...  \
0               0        0             0%                 0  ...   
1              61       48            48%                 0  ...   
2               0        0             0%                 0  ...   
3               0        0             0%                 0  ...   
4               1        1           100%                 0  ...   

   Accurate long balls  Yellow cards Red cards Fouls  Offsides Goals  \
0                    0             0         0     0         0     0   
1                  142             1         0    22         2     1   
2                    0             0         0     0         0     0   
3                    0             0         0     0         0     0   
4                    6             0         0     0         0     0   

  Headed goals Goals with right foot  Goals with left foot Hit woodwork  
0            0                     0                     0            0  
1            1                     0                     0            0  
2            0                     0                     0            0  
3            0                     0                     0            0  
4            0                     0                     0            0  

[5 rows x 40 columns]

In [124]:
dfMidfielder.head()

Player name  Appearances  Goals overview  Wins  Losses  Goals  \
0  Brenden Aaronson           14               1     4       7      1   
1       Tyler Adams           13               0     4       6      0   
2   Marc Albrighton          307              19   117     110     19   
3             Allan           52               0    20      23      0   
4     Charlie Allen            0               0     0       0      0   

   Goals per match  Headed goals Goals with right foot  Goals with left foot  \
0             0.07             0                     1                     0   
1            13.00             0                     0                     0   
2             0.06             2                    15                     2   
3            52.00             0                     0                     0   
4             0.00             0                     0                     0   

   ...  Interceptions  Clearances  Headed clearances  Recoveries Duels won  \
0  ...            1.0         5.0                1.0          68        77   
1  ...           16.0        11.0                2.0         104        75   
2  ...          292.0       204.0               60.0       1,437     1,128   
3  ...           42.0        59.0               23.0         378       273   
4  ...            0.0         0.0                0.0           0         0   

   Duels lost  Succesful 50/50s  Aerial battles won Aerial battles lost  \
0         106              19.0                 6.0                17.0   
1          62               2.0                10.0                 7.0   
2       1,346             336.0                84.0               233.0   
3         299              52.0                16.0                26.0   
4           0               0.0                 0.0                 0.0   

   Errors leading to goal  
0                     0.0  
1                     0.0  
2                     0.0  
3                     1.0  
4                     0.0  

[5 rows x 42 columns]

#### **1. Có bao nhiêu hàng và cột trong bộ dữ liệu?**
- Mỗi vị trí khác nhau sẽ có các thuộc tính dữ liệu khác nhau. Như vậy, mỗi bộ dữ liệu sẽ có các thông tin và đặc điểm riêng của nó.
- Bước đầu tiên, nhóm em đọc file dữ liệu đưa vào các DataFrame để xử lý:

In [125]:
# Midfielder
print("Số hàng (số cầu thủ) của tiền vệ: ", dfMidfielder.shape[0])
print("Số cột (số thuộc tính) của tiền vệ: ", dfMidfielder.shape[1])

Số hàng (số cầu thủ) của tiền vệ:  1808
Số cột (số thuộc tính) của tiền vệ:  42


In [126]:
# Defender
print("Số hàng (số cầu thủ) của hậu vệ: ", dfDefender.shape[0])
print("Số cột (số thuộc tính) của hậu vệ: ", dfDefender.shape[1])

Số hàng (số cầu thủ) của hậu vệ:  1673
Số cột (số thuộc tính) của hậu vệ:  40


In [127]:
# Keeper
print("Số hàng (số cầu thủ) của thủ môn: ", dfKeeper.shape[0])
print("Số cột (số thuộc tính) của thủ môn: ", dfKeeper.shape[1])

Số hàng (số cầu thủ) của thủ môn:  592
Số cột (số thuộc tính) của thủ môn:  25


In [128]:
# Forward
print("Số hàng (số cầu thủ) của tiền đạo: ", dfForward.shape[0])
print("Số cột (số thuộc tính) của tiền đạo: ", dfForward.shape[1])

Số hàng (số cầu thủ) của tiền đạo:  1062
Số cột (số thuộc tính) của tiền đạo:  31


#### **2. Mỗi cột và mỗi hàng trong các bộ dữ liệu có ý nghĩa gì?**
- Mỗi hàng trong bộ dữ liệu chỉ ra các chỉ số của cầu thủ trong tổng số các giải đấu ngoại hạng Anh mà họ đã tham gia trước đó, được trang chủ chính thức của Premier League tổng hợp. Có một số hàng bất thường chỉ chứa toàn giá trị là số 0, tức là cầu thủ đó chưa ra sân hoặc nằm ở vị trí dự bị.
- Mỗi cột ở mỗi bộ sẽ tương ứng với thuộc tính của cầu thủ trong bộ dữ liệu đó. Tất cả bộ dữ liệu có một vài thuộc tính giống nhau nhưng nhìn chung không nhiều, dưới đây là các thuộc tính đó:
  - **Player name**: Tên cầu thủ.
  - **Appearances**: Số lần ra sân của cầu thủ.
  - **Goals Overview**: Tổng số bàn thắng của các cầu thủ.
  - **Wins**: Tổng số trận thắng của cầu thủ.
  - **Losses**: Tổng số trận thua của cầu thủ.
  - **Yellow cards**: Tổng số thẻ vàng của cầu thủ.
  - **Red cards**: Tổng số thẻ đỏ của cầu thủ.
  - **Assists**: Tổng số đường kiến tạo.
  - **Passes**: Tổng số đường chuyền.
  - **Passes per match**: Số đường chuyền trung bình mỗi trận. 
- Ngoài ra, các bộ còn có đặc trưng riêng của mình, lần lượt được thể hiện trong các đoạn code sau đây:

---
**TIỀN XỬ LÝ DỮ LIỆU: LỌC BỎ CỘT TRÙNG**
- Mặc dù trong bộ dữ liệu theo quan sát thì không có thuộc tính nào trùng lặp, nhưng để chắc chắn thì nhóm em vẫn thực hiện việc xóa hết các cột trùng với nhau và in ra số lượng cột sau khi đã lọc trùng đó.

In [129]:
# Khi quan sát bộ dữ liệu, có một vài thuộc tính trùng, nhóm em xóa hết các thuộc tính 
# trùng đó và in ra số lượng cuối cùng thuộc tính:

# Lọc trùng của midfielder
temp = dfMidfielder.T.drop_duplicates().T
dfMidfielder = temp

# Lọc trùng của defender
temp = dfDefender.T.drop_duplicates().T
dfDefender = temp

# Lọc trùng của forward
temp = dfForward.T.drop_duplicates().T
dfForward = temp

# Lọc trùng của keeper
temp = dfKeeper.T.drop_duplicates().T
dfKeeper = temp

In [130]:
# Keeper
print("Số lượng thuộc tính sau khi lọc của thủ môn:", dfKeeper.columns.size)
dfKeeper.columns.to_numpy()

Số lượng thuộc tính sau khi lọc của thủ môn: 24


array(['Player name', 'Appearances', 'Clean sheets', 'Wins', 'Losses',
       'Saves', 'Penalties saved', 'Punches', 'High claims', 'Catches',
       'Sweeper clearances', 'Throw outs', 'Goal kicks', 'Goals conceded',
       'Errors leading to goal', 'Own goals', 'Yellow cards', 'Red cards',
       'Fouls', 'Goals', 'Assists', 'Passes', 'Passes per match',
       'Accurate long balls'], dtype=object)

- Trong tất cả các mùa giải, thủ môn có các thuộc tính như sau: 
  - **Clean sheets**: Tổng số trận giữ sạch lưới.
  - **Saves**: Tổng số lần cứu thua (cản phá cú sút).
  - **Penalties Saved**: Tổng số lần cản phá trên chấm phạt đền.
  - **Punches**: Tổng số cú đấm bóng.
  - **High Claims**: Tổng số cú bắt bóng từ những đường tạt bóng.
  - **Catches**: Tổng số cú bắt dính bóng từ những cú sút.
  - **Sweeper Clearances**: Tổng số tình huống cứu thua khi đối đầu trực tiếp với đối phương.
  - **Throw Outs**: Tổng số tình huống ném bóng lên.
  - **Goal Kicks**: Tổng số tình huống phát bóng lên.
  - **Goals Conceded**: Tổng số bàn thua.
  - **Errors leading to goal**: Tổng số lỗi dẫn đến bàn thua.
  - **Own goals**: Tổng số lần phản lưới nhà.
  - **Accurate long balls**: số đường chuyền dài chính xác.

In [131]:
# Defender
print("Số lượng thuộc tính sau khi lọc của hậu vệ:", dfDefender.columns.size)
dfDefender.columns.to_numpy()

Số lượng thuộc tính sau khi lọc của hậu vệ: 39


array(['Player name', 'Appearances', 'Goals Overview', 'Wins', 'Losses',
       'Clean sheets', 'Goals conceded', 'Tackles', 'Tackle success',
       'Last man tackles', 'Blocked shots', 'Interceptions', 'Clearances',
       'Headed clearance', 'Clearances off line', 'Recoveries',
       'Duels won', 'Duels lost', 'Successful 50/50s',
       'Aerial battles won', 'Aerial battles lost', 'Own goals',
       'Errors leading to goal', 'Assists', 'Passes', 'Passes per match',
       'Big chances created', 'Crosses', 'Cross accuracy',
       'Through balls', 'Accurate long balls', 'Yellow cards',
       'Red cards', 'Fouls', 'Offsides', 'Headed goals',
       'Goals with right foot', 'Goals with left foot', 'Hit woodwork'],
      dtype=object)

- Trong tất cả các mùa giải, hậu vệ có các thuộc tính như sau:   
  - Clean sheets $\Rightarrow$ Tổng số trận giữ sạch lưới.
  - Goals Conceded $\Rightarrow$ Tổng số bàn thua.
  - Tackle $\Rightarrow$ Tổng số lần chủ động đoạt bóng.
  - Tackle success $\Rightarrow$ Tổng số chủ động đoạt bóng bóng thành công.
  - Last man tackles $\Rightarrow$ Tổng số lần tranh chấp bóng thành công với cương vị là cầu thủ phòng thủ cuối cùng.
  - Blocked shots $\Rightarrow$ Tổng số tình huống ngăn cản cú sút của đối thủ.
  - Interceptions $\Rightarrow$ Tổng số tình huống đánh chặn.
  - Clearances $\Rightarrow$ Tổng số tình huống phá bóng lên bằng chân.
  - Headed Clearance $\Rightarrow$ Tổng số tình huống phá bóng lên bằng đầu.
  - Clearances off line $\Rightarrow$ Tổng số tình huống phá bóng lên ở gần vạch vôi.
  - Recoveries $\Rightarrow$ Tổng số tình huống chiếm lại quyền kiểm soát bóng.
  - Duels won $\Rightarrow$ Tổng số tình huống đối đầu thành công.
  - Duels lost $\Rightarrow$ Tổng số tình huống đối đầu thất bại.
  - Successful 50/50s $\Rightarrow$ Tổng số tình huống tranh chấp bóng sống thành công.
  - Aerial battles won $\Rightarrow$ Tổng số tình huống tranh chấp bóng bổng thành công.
  - Aerial battles lost $\Rightarrow$ Tổng số tình huống tranh chấp bóng bổng thất bại.
  - Own goals $\Rightarrow$ Tổng số lần phản lưới nhà.
  - Errors leading to goal $\Rightarrow$ Tổng số lỗi dẫn đến bàn thua.
  - Big chance created $\Rightarrow$ Tổng số những cơ hội thành bàn được tạo ra.
  - Crosses $\Rightarrow$ Tổng số đường tạt bóng.
  - Cross accuracy $\Rightarrow$ Tỉ lệ tạt bóng chính xác.
  - Through balls $\Rightarrow$ Tổng số đường chọc khe.
  - Accurate long balls $\Rightarrow$ Tổng số đường chuyển dài chính xác.
  - Offsides $\Rightarrow$ Tổng số lần rơi vào thế việt vị.
  - Headed goals $\Rightarrow$ Tổng số bàn thắng bằng đầu.
  - Goals with right foot $\Rightarrow$ Tổng số bàn thắng bằng chân phải.
  - Goals with left foot $\Rightarrow$ Tổng số bàn thắng bằng chân trái.
  - Hid woodword $\Rightarrow$ Tổng số cú sút chạm cột dọc hoặc xà ngang.

In [132]:
# Forward
print("Số lượng thuộc tính sau khi lọc của tiền đạo:", dfForward.columns.size)
dfForward.columns.to_numpy()

Số lượng thuộc tính sau khi lọc của tiền đạo: 30


array(['Player name', 'Appearances', 'Goals Overview', 'Wins', 'Losses',
       'Goals per match', 'Headed goals', 'Goals with right foot',
       'Goals with left foot', 'Penalties scored', 'Freekicks scored',
       'Shots', 'Shots on target', 'Shooting accuracy', 'Hit woodwork',
       'Big chances missed', 'Assists', 'Passes', 'Passes per match',
       'Big chances created', 'Crosses', 'Yellow cards', 'Red cards',
       'Fouls', 'Offsides', 'Tackles', 'Blocked shots', 'Interceptions',
       'Clearances', 'Headed clearance'], dtype=object)

- Trong tất cả các mùa giải, tiền đạo có các thuộc tính như sau: 
  - Goals per match $\Rightarrow$ Số bàn thắng trung bình mỗi trận.
  - Headed goals $\Rightarrow$ Tổng số bàn thắng bằng đầu.
  - Goals with right foot $\Rightarrow$ Tổng số bàn thắng bằng chân phải.
  - Goals with left foot $\Rightarrow$ Tổng số bàn thắng bằng chân trái.
  - Penalties scored $\Rightarrow$ Tổng số bàn thắng ghi từ chấm phạt đền.
  - Freekicks scored $\Rightarrow$ Tổng số bàn thắng ghi từ tình huống đá phạt trực tiếp.
  - Shots $\Rightarrow$ Tổng số cú sút.
  - Shots on target $\Rightarrow$ Tổng số cú sút trúng đích.
  - Shooting accuracy $\Rightarrow$ Tỉ lệ chính xác của những cú sút.
  - Hit woodwork $\Rightarrow$ Tổng số cú sút chạm cột dọc hoặc xà ngang.
  - Big chances missed $\Rightarrow$ Tổng những cơ hội thành bàn bị bỏ lỡ.
  - Big chance created $\Rightarrow$ Tổng số những cơ hội thành bàn được tạo ra.
  - Crosses $\Rightarrow$ Tổng số đường tạt bóng.
  - Offsides $\Rightarrow$ Tổng số lần rơi vào thế việt vị.
  - Tackle $\Rightarrow$ Tổng số lần chủ động đoạt bóng.
  - Blocked shots $\Rightarrow$ Tổng số tình huống ngăn cản cú sút của đối thủ.
  - Interceptions $\Rightarrow$ Tổng số tình huống đánh chặn.
  - Clearances $\Rightarrow$ Tổng số tình huống phá bóng lên bằng chân.
  - Headed Clearance $\Rightarrow$ Tổng số tình huống phá bóng lên bằng đầu.

In [133]:
# Midfielder
print("Số lượng thuộc tính sau khi lọc của tiền vệ:", dfMidfielder.columns.size)
dfMidfielder.columns.to_numpy()

Số lượng thuộc tính sau khi lọc của tiền vệ: 41


array(['Player name', 'Appearances', 'Goals overview', 'Wins', 'Losses',
       'Goals per match', 'Headed goals', 'Goals with right foot',
       'Goals with left foot', 'Penalties scored', 'Freekicks scored',
       'Shots', 'Shots on target', 'Shooting accuracy', 'Hit woodwork',
       'Big chances missed', 'Assists', 'Passes', 'Passes per match',
       'Big chances created', 'Crosses', 'Cross accuracy',
       'Through balls', 'Accurate long balls', 'Yellow cards',
       'Red cards', 'Fouls', 'Offsides', 'Tackles', 'Tackle success',
       'Blocked shots', 'Interceptions', 'Clearances',
       'Headed clearances', 'Recoveries', 'Duels won', 'Duels lost',
       'Succesful 50/50s', 'Aerial battles won', 'Aerial battles lost',
       'Errors leading to goal'], dtype=object)

- Trong tất cả các mùa giải, tiền vệ có các thuộc tính như sau: 
  - Goals per match $\Rightarrow$ Số bàn thắng trung bình mỗi trận.
  - Headed goals $\Rightarrow$ Tổng số bàn thắng bằng đầu.
  - Goals with right foot $\Rightarrow$ Tổng số bàn thắng bằng chân phải.
  - Goals with left foot $\Rightarrow$ Tổng số bàn thắng bằng chân trái.
  - Penalties scored $\Rightarrow$ Tổng số bàn thắng ghi từ chấm phạt đền.
  - Freekicks scored $\Rightarrow$ Tổng số bàn thắng ghi từ tình huống đá phạt trực tiếp.
  - Shots $\Rightarrow$ Tổng số cú sút.
  - Shots on target $\Rightarrow$ Tổng số cú sút trúng đích.
  - Shooting accuracy %$\Rightarrow$ Tỉ lệ chính xác của những cú sút.
  - Hit woodwork $\Rightarrow$ Tổng số cú sút chạm cột dọc hoặc xà ngang.
  - Big chances missed $\Rightarrow$ Tổng những cơ hội thành bàn bị bỏ lỡ.
  - Crosses $\Rightarrow$ Tổng số đường tạt bóng.
  - Cross accuracy $\Rightarrow$ Tỉ lệ tạt bóng chính xác.
  - Through balls $\Rightarrow$ Tổng số đường chọc khe.
  - Accurate long balls $\Rightarrow$ Tổng số đường chuyển dài chính xác.
  - Offsides $\Rightarrow$ Tổng số lần rơi vào thế việt vị.
  - Tackle $\Rightarrow$ Tổng số lần chủ động đoạt bóng.
  - Blocked shots $\Rightarrow$ Tổng số tình huống ngăn cản cú sút của đối thủ.
  - Interceptions $\Rightarrow$ Tổng số tình huống đánh chặn.
  - Clearances $\Rightarrow$ Tổng số tình huống phá bóng lên bằng chân.
  - Headed Clearance $\Rightarrow$ Tổng số tình huống phá bóng lên bằng đầu.
  - Recoveries $\Rightarrow$ Tổng số tình huống chiếm lại quyền kiểm soát bóng.
  - Duels won $\Rightarrow$ Tổng số tình huống đối đầu thành công.
  - Duels lost $\Rightarrow$ Tổng số tình huống đối đầu thất bại.
  - Successful 50/50s $\Rightarrow$ Tổng số tình huống tranh chấp bóng sống thành công.
  - Aerial battles won $\Rightarrow$ Tổng số tình huống tranh chấp bóng bổng thành công.
  - Aerial battles lost $\Rightarrow$ Tổng số tình huống tranh chấp bóng bổng thất bại.
  - Own goals $\Rightarrow$ Tổng số lần phản lưới nhà.
  - Errors leading to goal $\Rightarrow$ Tổng số lỗi dẫn đến bàn thua.

- Ở bộ dữ liệu nào cũng có thuộc tính suy diễn (có thể tính toán, suy ra từ các thuộc tính khác), nhóm em xóa các thuộc tính đó để có bộ dữ liệu đạt chuẩn hơn.
  - Thuộc tính suy diễn của thủ môn: Passes per match (= Passes / Appearances)
  - Thuộc tính suy diễn của hậu vệ: Passes per match (= Passes / Appearances)
  - Thuộc tính suy diễn của tiền vệ: Goals per match (= Goals / Appearances), Passes per match (= Passes / Appearances),
Shooting Accuracy (Shots on target / Shots)
  - Thuộc tính suy diễn của tiền đạo: Goals per match (= Goals / Appearances), Passes per match (= Passes / Appearances),
Shooting Accuracy (Shots on target / Shots)

---
**TIỀN XỬ LÝ DỮ LIỆU: BỎ THUỘC TÍNH SUY DIỄN, THUỘC TÍNH KHÔNG CẦN THIẾT:**

In [134]:
# CODE HERE - Dùng drop column - Thủ môn, trung vệ,...
dfKeeper.drop('Passes per match', axis = 1, inplace = True)
dfDefender.drop('Passes per match', axis = 1, inplace = True)
dfMidfielder.drop('Goals per match', axis = 1, inplace = True)
dfMidfielder.drop('Passes per match', axis = 1, inplace = True)
dfMidfielder.drop('Shooting accuracy', axis = 1, inplace = True)
dfForward.drop('Shooting accuracy', axis = 1, inplace = True)
dfForward.drop('Passes per match', axis = 1, inplace = True)
dfForward.drop('Goals per match', axis = 1, inplace = True)

In [135]:
a = dfMidfielder.columns.to_list()
b = dfForward.columns.to_list()
c = dfDefender.columns.to_list()

e = [ i for i in a if i in b and i in c]
print(e)

['Player name', 'Appearances', 'Wins', 'Losses', 'Headed goals', 'Goals with right foot', 'Goals with left foot', 'Hit woodwork', 'Assists', 'Passes', 'Big chances created', 'Crosses', 'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Tackles', 'Blocked shots', 'Interceptions', 'Clearances']


#### **3. Có sự lặp lại của các hàng không? Nếu có thì hãy chỉ ra sự trùng lắp đó.**
- Bộ dữ liệu được lấy từ 2 mùa giải gần nhất để đánh giá nên sẽ có một số cầu thủ bị trùng lắp không hoàn toàn hoặc bị trùng lắp hoàn toàn:
- **Sự trùng lắp hoàn toàn:** Các dòng giống nhau hoàn toàn, chỉ số và tên không thay đổi.
- **Sự trùng lắp không hoàn toàn**: Sự trùng lắp về tên, các thuộc tính khác hoàn toàn (hoặc giống hoàn toàn). Lý giải cho việc đó là có một số cầu thủ năm trước sẽ đá lại cho năm sau làm cho thuộc tính thay đổi.
- Vì dữ liệu theo hướng thực tế, cập nhật mỗi năm nên nhóm em chỉ lấy các cầu thủ xuất hiện đầu tiên (vì các thuộc tính mới nhất nằm ở đầu tiên vì phần tử đầu tiên được lấy từ bộ dữ liệu năm mới nhất), các cầu thủ trùng phía sau bị loại bỏ hết trong bộ dữ liệu:

In [136]:
# Đếm số giá trị trùng.
num_duplicate_data_Defender = dfDefender.duplicated().sum()
print("Số giá trị bị trùng lắp của hậu vệ: ", num_duplicate_data_Defender)

num_duplicate_data_Forward = dfForward.duplicated().sum()
print("Số giá trị bị trùng lắp của tiền đạo: ", num_duplicate_data_Forward)

num_duplicate_data_Keeper = dfKeeper.duplicated().sum()
print("Số giá trị bị trùng lắp của thủ môn: ", num_duplicate_data_Keeper)

num_duplicate_data_Midfielder = dfMidfielder.duplicated().sum()
print("Số giá trị bị trùng lắp của tiền vệ: ", num_duplicate_data_Midfielder)

Số giá trị bị trùng lắp của hậu vệ:  1335
Số giá trị bị trùng lắp của tiền đạo:  850
Số giá trị bị trùng lắp của thủ môn:  475
Số giá trị bị trùng lắp của tiền vệ:  1444


---
**TIỀN XỬ LÝ DỮ LIỆU: XÓA CÁC HÀNG TRÙNG:**

In [137]:
# Loại bỏ kiểu dữ liệu trùng lắp hoàn toàn trước:
# Tui tưởng trùng tên là loại th chứ trùng lắp hoàn toàn với k hoàn toàn làm gì á .-.
# CODE HERE - drop duplicate, đếm số dòng unique (dùng nunique) cho 4 vị trí.
dfDefender.drop_duplicates(inplace = True)
print(dfDefender.shape[0])

dfMidfielder.drop_duplicates(inplace = True)
print(dfMidfielder.shape[0])

dfForward.drop_duplicates(inplace = True)
print(dfForward.shape[0])

dfKeeper.drop_duplicates(inplace = True)
print(dfKeeper.shape[0])

338
364
212
117


- Vậy với vị trí:
  - Hậu vệ, ta có: 338 cầu thủ.
  - Tiền vệ, ta có: 364 cầu thủ.
  - Tiền đạo, ta có: 212 cầu thủ.
  - Thủ môn, ta có: 117 cầu thủ.

#### **4. Kiểu dữ liệu của mỗi cột là gì?**
- Để xử lý được cho các bước tiếp theo, nhóm em xác định các kiểu dữ liệu của mỗi cột dựa vào các mô tả thuộc tính ở trên, bắt đầu ở thuộc tính chung của các bộ (kiểu dữ liệu mong muốn):
  - **Player name**:              `string`
  - **Appearances**:              `int`
  - **Goals Overview**:           `int`
  - **Wins**:                     `int`
  - **Losses**:                   `int`
  - **Yellow cards**:             `int`
  - **Red cards**:                `int`
  - **Assists**:                  `int`
  - **Passes**:                   `int`
- Ứng với mỗi vị trí, các thuộc tính ứng với kiểu dữ liệu sau đây:


KIỂU DỮ LIỆU CỦA THỦ MÔN
- **Clean sheets**           `int`
- **Saves**                  `int`
- **Penalties Saved**        `int`    
- **Punches**                `int`
- **High Claims**            `int`
- **Catches**                `int`
- **Sweeper Clearances**     `int`        
- **Throw Outs**             `int`
- **Goal Kicks**             `int`
- **Goals Conceded**         `int`    
- **Errors leading to goal** `int`            
- **Own goals**              `int`
- **Accurate long balls**    `int`                                                


KIỂU DỮ LIỆU CỦA TIỀN VỆ
- **Headed goals**              `int`        
- **Goals with right foot**     `int`
- **Goals with left foot**      `int`
- **Penalties scored**          `int`
- **Freekicks scored**          `int`
- **Shots**                     `int`
- **Shots on target**           `int`
- **Shooting accuracy %**       `float`
- **Hit woodwork**              `int`
- **Big chances missed**        `int`
- **Assists**                   `int`
- **Big chances created**       `int`
- **Crosses**                   `int`
- **Cross accuracy %**          `int`
- **Through balls**             `int`
- **Accurate long balls**       `int`
- **Offsides**                  `int`
- **Tackles**                   `int`
- **Tackle success %**          `int`
- **Blocked shots**             `int`
- **Interceptions**             `int`
- **Clearances**                `int`
- **Headed clearances**         `int`
- **Recoveries**                `int`
- **Duels won**                 `int`
- **Duels lost**                `int`
- **Succesful 50/50s**          `int`
- **Aerial battles won**        `int`
- **Aerial battles lost**       `int`
- **Errors leading to goal**    `int`

KIỂU DỮ LIỆU CỦA HẬU VỆ
- **Clean sheets**              `int`
- **Goal Conceded**             `int`
- **Tackles**                   `int`
- **Tackle success %**          `int`
- **Last man tackles**          `int`
- **Blocked shots**             `int`
- **Interceptions**             `int`
- **Clearances**                `int`
- **Headed clearances**         `int`
- **Recoveries**                `int`
- **Duels won**                 `int`
- **Duels lost**                `int`
- **Succesful 50/50s**          `int`
- **Aerial battles won**        `int`
- **Aerial battles lost**       `int`
- **Own goals**                 `int`
- **Errors leading to goal**    `int`
- **Assists**                   `int`
- **Big chances created**       `int`
- **Crosses**                   `int`
- **Cross accuracy %**          `int`
- **Through balls**             `int`
- **Accurate long balls**       `int`
- **Offsides**                  `int`
- **Headed goals**              `int`        
- **Goals with right foot**     `int`
- **Goals with left foot**      `int`
- **Hit woodwork**              `int`

KIỂU DỮ LIỆU CỦA TIỀN ĐẠO
- **Headed goals**              `int`        
- **Goals with right foot**     `int`
- **Goals with left foot**      `int`
- **Penalties scored**          `int`
- **Freekicks scored**          `int`
- **Shots**                     `int`
- **Shots on target**           `int`
- **Shooting accuracy %**       `float`
- **Hit woodwork**              `int`
- **Big chances missed**        `int`
- **Assists**                   `int`
- **Big chances created**       `int`
- **Crosses**                   `int`
- **Cross accuracy %**          `int`
- **Through balls**             `int`
- **Accurate long balls**       `int`
- **Offsides**                  `int`
- **Tackles**                   `int`
- **Blocked shots**             `int`
- **Interceptions**             `int`
- **Clearances**                `int`
- **Headed clearances**         `int`

In [138]:
# CODE HERE - IN KIỂU DỮ LIỆU HIỆN TẠI...
Forward_types = dfForward.dtypes
Forward_types

Player name              object
Appearances              object
Goals Overview           object
Wins                     object
Losses                   object
Headed goals             object
Goals with right foot    object
Goals with left foot     object
Penalties scored         object
Freekicks scored         object
Shots                    object
Shots on target          object
Hit woodwork             object
Big chances missed       object
Assists                  object
Passes                   object
Big chances created      object
Crosses                  object
Yellow cards             object
Red cards                object
Fouls                    object
Offsides                 object
Tackles                  object
Blocked shots            object
Interceptions            object
Clearances               object
Headed clearance         object
dtype: object

In [139]:
Defender_types = dfDefender.dtypes
Defender_types

Player name               object
Appearances               object
Goals Overview            object
Wins                      object
Losses                    object
Clean sheets              object
Goals conceded            object
Tackles                   object
Tackle success            object
Last man tackles          object
Blocked shots             object
Interceptions             object
Clearances                object
Headed clearance          object
Clearances off line       object
Recoveries                object
Duels won                 object
Duels lost                object
Successful 50/50s         object
Aerial battles won        object
Aerial battles lost       object
Own goals                 object
Errors leading to goal    object
Assists                   object
Passes                    object
Big chances created       object
Crosses                   object
Cross accuracy            object
Through balls             object
Accurate long balls       object
Yellow car

In [140]:
Keeper_types = dfKeeper.dtypes
Keeper_types


Player name               object
Appearances               object
Clean sheets              object
Wins                      object
Losses                    object
Saves                     object
Penalties saved           object
Punches                   object
High claims               object
Catches                   object
Sweeper clearances        object
Throw outs                object
Goal kicks                object
Goals conceded            object
Errors leading to goal    object
Own goals                 object
Yellow cards              object
Red cards                 object
Fouls                     object
Goals                     object
Assists                   object
Passes                    object
Accurate long balls       object
dtype: object

In [141]:
MidFielder_types = dfMidfielder.dtypes
MidFielder_types

Player name               object
Appearances               object
Goals overview            object
Wins                      object
Losses                    object
Headed goals              object
Goals with right foot     object
Goals with left foot      object
Penalties scored          object
Freekicks scored          object
Shots                     object
Shots on target           object
Hit woodwork              object
Big chances missed        object
Assists                   object
Passes                    object
Big chances created       object
Crosses                   object
Cross accuracy            object
Through balls             object
Accurate long balls       object
Yellow cards              object
Red cards                 object
Fouls                     object
Offsides                  object
Tackles                   object
Tackle success            object
Blocked shots             object
Interceptions             object
Clearances                object
Headed cle

- Kiểu dữ liệu mong muốn để xử lý vị trí mỗi cầu thủ (này để Nguyên xử lý)

#### **5. Thông kế chi tiết của DataFrame**

In [142]:
dfForward.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212 entries, 0 to 946
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Player name            212 non-null    object
 1   Appearances            212 non-null    object
 2   Goals Overview         212 non-null    object
 3   Wins                   212 non-null    object
 4   Losses                 212 non-null    object
 5   Headed goals           212 non-null    object
 6   Goals with right foot  212 non-null    object
 7   Goals with left foot   212 non-null    object
 8   Penalties scored       212 non-null    object
 9   Freekicks scored       212 non-null    object
 10  Shots                  212 non-null    object
 11  Shots on target        212 non-null    object
 12  Hit woodwork           212 non-null    object
 13  Big chances missed     212 non-null    object
 14  Assists                212 non-null    object
 15  Passes                 

In [143]:
dfMidfielder.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 364 entries, 0 to 1622
Data columns (total 38 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Player name             364 non-null    object
 1   Appearances             364 non-null    object
 2   Goals overview          364 non-null    object
 3   Wins                    364 non-null    object
 4   Losses                  364 non-null    object
 5   Headed goals            364 non-null    object
 6   Goals with right foot   364 non-null    object
 7   Goals with left foot    364 non-null    object
 8   Penalties scored        364 non-null    object
 9   Freekicks scored        364 non-null    object
 10  Shots                   364 non-null    object
 11  Shots on target         364 non-null    object
 12  Hit woodwork            364 non-null    object
 13  Big chances missed      364 non-null    object
 14  Assists                 364 non-null    object
 15  Passe

In [144]:
dfKeeper.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 117 entries, 0 to 119
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Player name             117 non-null    object
 1   Appearances             117 non-null    object
 2   Clean sheets            117 non-null    object
 3   Wins                    117 non-null    object
 4   Losses                  117 non-null    object
 5   Saves                   117 non-null    object
 6   Penalties saved         117 non-null    object
 7   Punches                 117 non-null    object
 8   High claims             117 non-null    object
 9   Catches                 117 non-null    object
 10  Sweeper clearances      117 non-null    object
 11  Throw outs              117 non-null    object
 12  Goal kicks              117 non-null    object
 13  Goals conceded          117 non-null    object
 14  Errors leading to goal  117 non-null    object
 15  Own go

In [145]:
dfDefender.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 338 entries, 0 to 1505
Data columns (total 38 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Player name             338 non-null    object
 1   Appearances             338 non-null    object
 2   Goals Overview          338 non-null    object
 3   Wins                    338 non-null    object
 4   Losses                  338 non-null    object
 5   Clean sheets            338 non-null    object
 6   Goals conceded          338 non-null    object
 7   Tackles                 338 non-null    object
 8   Tackle success          338 non-null    object
 9   Last man tackles        338 non-null    object
 10  Blocked shots           338 non-null    object
 11  Interceptions           338 non-null    object
 12  Clearances              338 non-null    object
 13  Headed clearance        338 non-null    object
 14  Clearances off line     338 non-null    object
 15  Recov

#### **6. Độ quan trọng của thuộc tính.**

Ta sẽ giữ lại các thuộc tính này ở các vị trí trừ thủ môn

['Player name', 'Appearances', 'Goals Overview', 'Wins', 'Losses', 'Headed goals', 'Goals with right foot', 'Goals with left foot', 'Hit woodwork', 'Assists', 'Passes', 'Crosses', 'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Tackles', 'Blocked shots', 'Interceptions', 'Clearances']

- Tiền đạo bỏ: Penalty scored, Freekick scored,
- Tiền vệ bỏ: Penalty scored, Freekick scored, Big chance missed, Tackle success, Head clearance, Recoveries, Duels lost, Duels won, Successful 50/50s, Aerial battles win, Aerial battles lost, Errors leading to goal,
- Hậu vệ bỏ: Tackle success, Last man tackle, Recoveries, Duels lost, Duels won, Successful 50/50s, Aerial battles win, Aerial battles lost, Own goals, Cross accuracy, Accurate through balls.
- Thủ môn bỏ: Clean sheets, Saves, Penalties saved, Throw outs, Goal kicks, Goals conceded, Errors leading to goal, Own goals, Assists, Passes, Accurate long balls

In [146]:
dfForward.drop("Penalty scored", inplace=True, axis=1, errors='ignore')
dfForward.drop("Freekick scored", inplace=True, axis=1, errors='ignore')

dfMidfielder.drop("Penalty scored", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Freekick scored", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Big chance missed", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Tackle success", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Head clearance", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Recoveries", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Duels lost", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Duels won", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Successful 50/50s", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Aerial battles win", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Aerial battles lost", inplace=True, axis=1, errors='ignore')
dfMidfielder.drop("Errors leading to goal", inplace=True, axis=1, errors='ignore')

dfDefender.drop("Tackle success", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Last man tackle", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Recoveries", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Duels lost", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Duels won", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Successful 50/50s", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Aerial battles win", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Aerial battles lost", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Own goals", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Cross accuracy", inplace=True, axis=1, errors='ignore')
dfDefender.drop("Accurate through balls", inplace=True, axis=1, errors='ignore')

dfKeeper.drop("Clean sheets", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Saves", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Penalties saved", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Throw outs", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Goal kicks", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Goals conceded", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Errors leading to goal", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Own goals", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Assists", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Passes", inplace=True, axis=1, errors='ignore')
dfKeeper.drop("Accurate long balls", inplace=True, axis=1, errors='ignore')


#### **7. Phân loại thuộc tính.**
- Các thuộc tính đều là thuộc tính numeric

#### Task 3: Data Preprocessing
- Loại bỏ các kí hiệu, các kí tự đặc biệt '/n', '/t', '%',... trong các cột dữ liệu.
- Chuyển đổi kiểu dữ liệu sao cho phù hợp với bài và để tính toán về sau (kiểu dữ liệu của Pandas).
- In ra những dòng trống, rỗng, NaN, toàn bộ là 0 và thuộc tính tương ứng của nó để bổ sung sau.
- Điền giá trị thiếu bằng phương pháp constant (điền giá trị 0), mean, median vào giá trị thiếu cho cột numeric, và mode cho cột categorical, làm tròn ở 2 chữ số thập phân.

In [147]:
(dfMidfielder['Appearances'] == 0).value_counts()

False    233
True     131
Name: Appearances, dtype: int64

In [148]:
(dfDefender['Appearances'] == 0).value_counts()

False    210
True     128
Name: Appearances, dtype: int64

In [149]:
(dfForward['Appearances'] == 0).value_counts()

False    158
True      54
Name: Appearances, dtype: int64

In [150]:
(dfKeeper['Appearances'] == 0).value_counts()

True     60
False    57
Name: Appearances, dtype: int64